scripts/builtin/cspline.dml

```diff
@@ -0,0 +1,63 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+#
+# THIS SCRIPT SOLVES CUBIC SPLINE INTERPOLATION
+#
+# INPUT PARAMETERS:
+# --------------------------------------------------------------------------------------------
+# NAME  TYPE           DEFAULT   MEANING
+# --------------------------------------------------------------------------------------------
+# X     Matrix[Double]  ---      1-column matrix of x values knots
+# Y     Matrix[Double]  ---      1-column matrix of corresponding y values knots
+# inp_x Double          ---      the given input x, for which the cspline will find predicted y.
+# Log   String          " "      Location to store iteration-specific variables for monitoring and debugging purposes
+#
+# tol   Double          0.000001 Tolerance (epsilon); conjugate graduent procedure terminates early if
+#                                L2 norm of the beta-residual is less than tolerance * its initial norm
+# maxi  Int             0        Maximum number of conjugate gradient iterations, 0 = no maximum
+# --------------------------------------------------------------------------------------------
+# OUTPUT: 
+# pred_Y Matrix[Double] ---      Predicted value
+# K      Matrix[Double] ---      Matrix of k parameters
```

Comment:
Does K contain all information to make predictions on new data? if so maybe if you have time make a predict builtin

Reply:
K does not contain all information needed to make predictions. For a prediction, we need X and Y as well. However, one could make interpSpline() function as builtin, since it is a duplicate code in both csplineDS and csplineCG which would need X, Y and K as arguments. It does make sense, but we do not have time to do it right now. Maybe you can generate a new ticket for it, so we could finish it later. Thank you for the suggestion :)

Reply:
i am assuming that a new predict function would take a new X dataset and the content of K, and be able to make a new pred_y on unseen data.
I we would need Y to make a prediction, then it would be the same as saying give me a dataset and "what is in it (labels)", and i will tell you "what is in it".

again this is a task that is only bonus if possible.
for inspiration take a look at "pca.dml" and "pcaPredict.dml, there pca is used to "train" parameters, and pcaPredict is used on unseen data.

Reply:
With X and Y we meant sample points. Sorry for leaving it ambiguous. For the new point we would do the following:

https://en.wikipedia.org/wiki/Spline_interpolation#Algorithm_to_find_the_interpolating_cubic_spline

So K will be calculated from sample points using cspline builtin function and for each new x, we would need already calculated K, sample points X, Y and new x.

It is too late to include it in this PR since it is a submission for the DIA practical, but I could do it in the following weeks in a new PR.

Reply:
In your case the submission deadline shall not stop you from improving upon your implementation that you clearly did not hand in too late. If you think the remaining issues can be fixed within the next two weeks I suggest to leave the PR open and add [WIP] at the beginning of the title to mark it as being worked on.

Reply:
I think the submission is fine, but to make it practical to use we need the predict. This could be a new JIRA Task.